source/llvm/LLVMModelDataSymbols.cpp

```diff
@@ -1625,17 +1625,42 @@ size_t LLVMModelDataSymbols::getEventBufferSize(size_t eventId) const
 
 bool LLVMModelDataSymbols::isNamedSpeciesReference(const std::string& id) const
 {
-    return namedSpeciesReferenceInfo.find(id) != namedSpeciesReferenceInfo.end();
+    if (namedSpeciesReferenceInfo.find(id) != namedSpeciesReferenceInfo.end()) {
+        return true;
+    }
+    if (id.find(":") != string::npos) {
+        //LS: This whole section is for when we ask for rateOf(specid).
+        string rxnid = id.substr(0, id.find(":"));
+        string specid = id.substr(id.find(":") + 1, id.length());
+        if (getReactionIndex(rxnid) != -1 && getFloatingSpeciesIndex(specid) != -1) {
+            return true;
+        }
+    }
+    return false;
 }
 
 const LLVMModelDataSymbols::SpeciesReferenceInfo&
-    LLVMModelDataSymbols::getNamedSpeciesReferenceInfo(const std::string& id) const
+    LLVMModelDataSymbols::getNamedSpeciesReferenceInfo(const std::string& id)
 {
     StringRefInfoMap::const_iterator i = namedSpeciesReferenceInfo.find(id);
     if (i != namedSpeciesReferenceInfo.end())
     {
         return i->second;
     }
+    else if (id.find(":") != string::npos) {
+        //LS: This whole section is for when we ask for rateOf(specid).
+        string rxnid = id.substr(0, id.find(":"));
+        string specid = id.substr(id.find(":") + 1, id.length());
+        int rxnIdx = getReactionIndex(rxnid);
+        int speciesIdx = getFloatingSpeciesIndex(specid);
+        if (rxnIdx == -1 || speciesIdx == -1) {
+            throw_llvm_exception(id + " is not a named SpeciesReference: '" + rxnid + "' and '" + specid + "' are not a valid combination of reaction and species.");
+        }
+        SpeciesReferenceInfo info =
+        { static_cast<uint>(speciesIdx), static_cast<uint>(rxnIdx), Product, rxnid };
+        namedSpeciesReferenceInfo[id] = info;
+        return namedSpeciesReferenceInfo[id];
+    }
```

Comment:
If we want to add these species reference info to the namedSpeciesReferenceInfo, I am wondering why don't we instead put these part somewhere where the model info is loaded?

Reply:
Because 99.999% of the time, they would be useless.  The only time we need them is in this very specific case when you have a model with a rateOf() for a floating species.  I could create them anyway (and maybe someone would find them helpful?) but I didn't want to add a lot of baggage for a little-used feature.

Reply:
Not at all. All I'm suggesting is that not adding it into namedSpeciesReferenceInfo so that each time this function is called, it is recalculated but not added.

Reply:
I can't do that, unfortunately:  the function returns a *reference* to an 'info' object, so I have to store it somewhere.  This was the 'it fails in release but works in debug' bug I ran into earlier.

Reply:
Then I would say it looks good the way it is.